src/hash.cpp

```diff
@@ -0,0 +1,156 @@
+    outDigest = utility::toHex(value, length);
+}
+
+static void _hash(DigestTypes digestType, const std::vector<uint8_t> &message, std::vector<uint8_t> &outDigest) {
+    uint8_t value[Hash::lengthBytes[digestType]];
+    unsigned int length;    
+    _hash(digestType, &message[0], message.size(), value, length);
+    outDigest.insert(outDigest.begin(), value, value + length);
+    outDigest.resize(length);
+}
+
+static const std::string SHA256 = "sha256";
+static const std::string SHA512 = "sha512";
+
+void sha256(const uint8_t* message, const int &length, uint8_t* outDigest, unsigned int &outLength) {
+    _hash(DigestTypes::SHA256, message, length, outDigest, outLength);
+}
+
+void sha256(const std::string &message, std::string &outDigest) {
+    _hash(DigestTypes::SHA256, message, outDigest);
+}
+
+void sha256(const std::vector<uint8_t> &message, std::vector<uint8_t> &outDigest) {
+    _hash(DigestTypes::SHA256, message, outDigest);
+}
+
+void sha512(const uint8_t* message, const int &length, uint8_t* outDigest, unsigned int &outLength) {
+    _hash(DigestTypes::SHA512, message, length, outDigest, outLength);
+}
+
+void sha512(const std::string &message, std::string &outDigest) {
+    _hash(DigestTypes::SHA512, message, outDigest);
+}
+
+void sha512(const std::vector<uint8_t> &message, std::vector<uint8_t> &outDigest) {
+    _hash(DigestTypes::SHA512, message, outDigest);
+}
+
+std::map<DigestTypes, int> Hash::lengthBytes = {
+    { DigestTypes::SHA256, 256 / 8 },
+    { DigestTypes::SHA512, 512 / 8 }
+};
+
+Hash::Hash(const DigestTypes _digestType) : _digestType(_digestType) {
+    _digestCtx = _initCtx(_digestType);
+}
+
+Hash Hash::sha256 () {
+    return Hash{DigestTypes::SHA256};
+}
+
+Hash Hash::sha512 () {
+    return Hash{DigestTypes::SHA512};
+}
+
+void Hash::digest (std::string &outValue) {
+    unsigned int length;
+    uint8_t value[getDigestLengthBytes()];
+    digest(value, length);
+    outValue = utility::toHex(value, length);
+}
+
+void Hash::digest (std::vector<uint8_t> &outValue) {
+    unsigned int length;
+    uint8_t value[getDigestLengthBytes()];
+    digest(value, length);
+    outValue.insert(outValue.begin(), value, value + length);
+    outValue.resize(length);
+}
+
+void Hash::digest (uint8_t* outValue, unsigned int &length) {
+    _finalizeDigest(_digestCtx, outValue, length);
+    _finalized = true;
+}
+
+unsigned int Hash::getDigestLengthBytes () {
+    return Hash::lengthBytes[_digestType];
+}
+
+Hash& Hash::update (const std::string &chunk) {
+    return update(reinterpret_cast<const uint8_t*>(chunk.c_str()), chunk.length());
+}
+
+Hash& Hash::update (const std::vector<uint8_t> &chunk) {
+    return update(&chunk[0], chunk.size());
+}
+
+Hash& Hash::update (const uint8_t* chunk, const int &length) {
+    if(_finalized) {
+        throw std::runtime_error("update method cannot be called after digest was called");
```

Comment:
Reconsider exception strategy. Do we throw runtime error anywhere else? I think this library mostly throws MoCoCrW error.

Reply:
done

Reply:
looks still like runtime_error to me? Or is there any rationale behind this.